File: scripts/icerik-hatti/_veri.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""ORTAK VERI ERISIMI — salt okuma + veri-tamligi kapisi (REC-163/168 recetesi).

NICIN AYRI MODUL: ayni kapi iki betikte KOPYA dursa, biri duzeltilip oteki unutulur ve
gunun birinde "iki betik ayni soruya iki cevap verir" haline duseriz. Tek yerde durur.
(2026-09-06 aksami tam bu oldu: fiyatsiz-ayrim.py kendi kopyasini tasiyordu, select=* duzeltmesi
ona girmemisti ve sayfalamasi SIRASIZDI — workflow curutmesi buldu. Kopyalar kaldirildi.)

⭐OLCULDU 2026-09-06 (dort ayri vaka, dordu de sessizdi):
1. PostgREST tek cagrida EN COK 1000 satir doner (Supabase projesinin max-rows AYARI; varsayilan
   1000, Dashboard/Management API ile degisebilir — depoda kaydi yok). `limit=2000` ISE YARAMAZ.
   product_prices 1044 satir -> 44 satir SESSIZCE dustu.
2. Kesin sayi alinamazsa denetimi ATLAMAK fail-open'dir: kapi tam gerektigi anda
   kendini kapatir. "Olcemedim" ile "temiz" ayni dala DUSMEZ.
3. Dongu tavani yoksa, sayfalama bozuldugunda (offset ilerlemezse) SONSUZ dongu.
4. SIRASIZ sayfalama (order= yok) satir atlar/tekrarlar ve toplam sayi YINE tutar — kesin sayi
   kapisi bunu GOREMEZ. Her sayfali okumada deterministik sira sarttir (varsayilan order=id).

⚠`count=exact` her istemcide ayni yerde kabul EDILMEZ. Burada ham urllib + `Prefer`
basligi kullaniliyor ve Content-Range olculuyor. supabase-js'te secenek zincirin sonunda
.select() ile istenirse YUTULUYOR (ALTYAPI olctu). Kural: "her yerde calisir" degil,
"istemcin nerede kabul ediyor, OLC".

SINAV (her degisiklikte): SAYFA_BOYU=100 ile ≤1000 satirlik tabloda da sayfalama yolu KOSAR;
sabotaj A (Prefer kopuk) -> cikis 1; sabotaj B (offset ilerlemesin) -> DONGU TAVANI, cikis 1.
Tablo tek sayfaya sigiyorsa sabotaj B bos sinavdir — sayfa boyunu kucult.
"""
from __future__ import annotations

import json
import os
import re
import sys
import urllib.request
from pathlib import Path


def env_oku() -> dict:
    yol = Path(os.environ.get("VENTHUB_ENV") or (Path.home() / "venthub-hvac" / ".env"))
    o = {}
    for satir in yol.read_text(encoding="utf-8").splitlines():
        if not satir or satir.startswith("#") or "=" not in satir:
            continue
        k, v = satir.split("=", 1)
        o[k.strip()] = v.strip().strip("\"'")
    return o


def baglan() -> tuple[str, dict]:
    """(URL, basliklar). Anon anahtar KABUL EDILMEZ — RLS altinda sessizce BOS doner
    ve bos veri "hic bosluk yok" gibi gorunur; en tehlikeli sahte yesil."""
    o = env_oku()
    U = o.get("SUPABASE_URL") or o.get("NEXT_PUBLIC_SUPABASE_URL")
    K = o.get("SUPABASE_SERVICE_ROLE_KEY")
    if not (U and K):
        raise SystemExit("⛔ SUPABASE_URL / SERVICE_ROLE_KEY yok — anon ile olculmez.")
    return U, {"apikey": K, "Authorization": "Bearer " + K}


def rest(U: str, h: dict, yol: str):
    istek = urllib.request.Request(f"{U}/rest/v1/{yol}", headers=h)
    with urllib.request.urlopen(istek) as y:
        return json.loads(y.read().decode("utf-8"))


def _sorgu_ekle(yol: str, ek: str) -> str:
    return yol + ("&" if "?" in yol else "?") + ek


def kesin_sayi(U: str, h: dict, yol: str) -> int:
    """`yol` bir tablo adi ("products") YA DA filtreli sorgu ("products?select=id&status=eq.active")
    olabilir; sayi AYNI filtreyle alinir (tablo sayisi ile filtreli cekim karsilastirilirsa kapi
    daima kirmizi olur — workflow bulgusu). select=* : 'id' kolonu olmayan tablo/gorunumlerde
    select=id 400 verir ve olcum "olculemedi"ye duser (2026-09-06 taramasinda 8 nesne)."""
    sorgu = yol if "?" in yol else f"{yol}?select=*"
    istek = urllib.request.Request(f"{U}/rest/v1/{_sorgu_ekle(sorgu, 'limit=1')}",
                                   headers={**h, "Prefer": "count=exact"})
    with urllib.request.urlopen(istek) as y:
        cr = y.headers.get("Content-Range") or ""
    son = cr.split("/")[-1] if "/" in cr else ""
    if not son.isdigit():
        raise SystemExit(f"⛔ OLCUM GUVENILIR DEGIL: {yol} icin kesin sayi alinamadi "
                         f"(Content-Range: {cr!r}). Cikti uretilmedi.")
    return int(son)


def tumunu_cek(U: str, h: dict, yol: str, tablo: str, sira: str = "id") -> list:
    """Sayfalar VE sayfalamanin dogru calistigini OLCER. Ikisi ayri sey.
    - kesin sayi AYNI filtreyle (yol) alinir;
    - `order=` yoksa `order=<sira>` eklenir (sirasiz sayfalama satir atlar, sayi yine tutar);
    - sayfa boyu SAYFA_BOYU ortam degiskeniyle kuculur (sinav icin), varsayilan 1000."""
    boy = int(os.environ.get("SAYFA_BOYU") or 1000)
    if boy < 1 or boy > 1000:
        raise SystemExit(f"⛔ SAYFA_BOYU {boy} gecersiz (1..1000).")
    if not re.search(r"(^|[?&])order=", yol):
        yol = _sorgu_ekle(yol, f"order={sira}")
    kesin = kesin_sayi(U, h, yol)
    tur_tavani = kesin // boy + 2
    top, bas, tur = [], 0, 0
    while True:
        tur += 1
        if tur > tur_tavani:
            raise SystemExit(f"⛔ DONGU TAVANI asildi: {tablo} — {tur} tur, beklenen en cok "
                             f"{tur_tavani}. Sayfalama bozuk; cikti uretilmedi.")
        parca = rest(U, h, _sorgu_ekle(yol, f"offset={bas}&limit={boy}"))
        if not isinstance(parca, list):
            raise SystemExit(f"⛔ BEKLENMEYEN CEVAP: {tablo} — liste degil: {str(parca)[:120]}")
        if not parca:
            break
        top += parca
        if len(parca) < boy:
            break
        bas += boy
    if len(top) != kesin:
        raise SystemExit(f"⛔ EKSIK VERI: {tablo} — cekilen {len(top)}, sunucu {kesin}. "
                         "Olcum GECERSIZ; cikti uretilmedi.")
    return top


def utf8_akis():
    for a in (sys.stdout, sys.stderr):
        try:
            a.reconfigure(encoding="utf-8", errors="replace")
        except Exception:
            pass

File: scripts/icerik-hatti/_veri.mjs
/**
 * ORTAK VERI ERISIMI (JS) — 1000 satir tavanina karsi sayfalama + veri-tamligi kapisi.
 *
 * `_veri.py`'nin JS esdegeri. NICIN AYRI DEGIL ORTAK: ayni kapi her betikte KOPYA dursa,
 * biri duzeltilip oteki unutulur (2026-09-06'da tam bu oldu — fiyatsiz-ayrim.py kendi
 * kopyasini tasiyordu ve `select=*` duzeltmesi ona girmemisti).
 *
 * ⭐OLCULDU (2026-09-06/07, filo notu `icerik-hatti-1000-satir-tavani-filo-notu-2026-09-06.md`):
 *  1. PostgREST tek cagrida en cok 1000 satir doner (Supabase `max_rows` AYARI; varsayilan 1000,
 *     Dashboard/Management API ile degisir). `limit=2000` ISE YARAMAZ — 2000 istendi, 1000 geldi.
 *  2. Kesin sayi alinamazsa denetimi ATLAMAK fail-open'dir: kapi tam gerektigi anda kapanir.
 *  3. Dongu tavani yoksa sayfalama bozulunca (offset ilerlemezse) SONSUZ dongu.
 *  4. SIRASIZ sayfalama satir atlar/tekrarlar ve toplam sayi YINE tutar -> `order=` sart.
 *  5. Sayi AYNI FILTREYLE alinmali; tablo toplami ile filtreli cekim karsilastirilirsa kapi
 *     daima kirmizi (ya da filtre unutulursa daima yesil) olur.
 *  6. `Number.isInteger` sart: PostgREST `*` donerse `parseInt('*')` = NaN, `typeof NaN === 'number'`
 *     GECER ve tavan NaN olur -> sonsuz dongu (workflow curutmesi buldu).
 *
 * SINAV: `SAYFA_BOYU=100` ile ≤1000 satirlik tabloda da sayfalama yolu KOSAR. Tablo tek sayfaya
 * sigiyorsa "offset ilerlemesin" sabotaji BOS SINAVDIR — sayfa boyunu kucult.
 */

/** `yol` = "products?select=id&brand=eq.SEAT" gibi TAM sorgu (tablo + filtre). */
function sorguEkle(yol, ek) {
  return yol + (yol.includes('?') ? '&' : '?') + ek
}

/** Sunucunun bildirdigi KESIN satir sayisi — AYNI filtreyle. Alinamazsa KIRMIZI. */
export async function kesinSayi(dbUrl, basliklar, yol) {
  const sorgu = yol.includes('?') ? yol : `${yol}?select=*`
  const cevap = await fetch(`${dbUrl}/rest/v1/${sorguEkle(sorgu, 'limit=1')}`, {
    headers: { ...basliklar, Prefer: 'count=exact' },
  })
  if (!cevap.ok) {
    throw new Error(`OLCUM GUVENILIR DEGIL: ${yol} sayim istegi HTTP ${cevap.status}`)
  }
  const aralik = cevap.headers.get('content-range') || ''
  const son = aralik.split('/').pop() || ''
  if (!/^\d+$/.test(son)) {
    throw new Error(
      `OLCUM GUVENILIR DEGIL: ${yol} icin kesin sayi alinamadi (Content-Range: ${JSON.stringify(aralik)}). ` +
      'Cikti uretilmedi.')
  }
  const n = Number(son)
  if (!Number.isInteger(n) || n < 0) {
    throw new Error(`OLCUM GUVENILIR DEGIL: ${yol} kesin sayi tamsayi degil (${son})`)
  }
  return n
}

/**
 * Sayfalar VE sayfalamanin dogru calistigini OLCER. Ikisi ayri sey.
 * Doner: satir dizisi. Herhangi bir tutarsizlikta THROW eder (cagiran cikis 1 vermeli).
 */
export async function tumSatirlar(dbUrl, basliklar, yol, secenek = {}) {
  const sira = secenek.sira || 'id'
  const boy = Number(secenek.boy || process.env.SAYFA_BOYU || 1000)
  if (!Number.isInteger(boy) || boy < 1 || boy > 1000) {
    throw new Error(`SAYFA_BOYU gecersiz: ${boy} (1..1000)`)
  }
  const sirali = /(^|[?&])order=/.test(yol) ? yol : sorguEkle(yol, `order=${sira}`)
  const kesin = await kesinSayi(dbUrl, basliklar, sirali)
  const turTavani = Math.floor(kesin / boy) + 2
  let top = []
  let bas = 0
  let tur = 0
  for (;;) {
    tur += 1
    if (tur > turTavani) {
      throw new Error(`DONGU TAVANI asildi: ${yol} — ${tur} tur, beklenen en cok ${turTavani}. ` +
        'Sayfalama bozuk; cikti uretilmedi.')
    }
    const cevap = await fetch(`${dbUrl}/rest/v1/${sorguEkle(sirali, `offset=${bas}&limit=${boy}`)}`,
      { headers: basliklar })
    if (!cevap.ok) throw new Error(`DB okuma hatasi ${cevap.status} (${yol})`)
    const parca = await cevap.json()
    if (!Array.isArray(parca)) {
      throw new Error(`BEKLENMEYEN CEVAP: ${yol} — liste degil: ${JSON.stringify(parca).slice(0, 120)}`)
    }
    if (parca.length === 0) break
    top = top.concat(parca)
    if (parca.length < boy) break
    bas += boy
  }
  if (top.length !== kesin) {
    throw new Error(`EKSIK VERI: ${yol} — cekilen ${top.length}, sunucu ${kesin}. ` +
      'Olcum GECERSIZ; cikti uretilmedi.')
  }
  return top
}
